### ISSUE TITLE
add batch axis to hk.Module

### ISSUE THREAD

**Issue by dyth** [deepmind/dm-haiku#315]
hi!

In flax, I could add a batch axis to the parameters of an arbitrary flax module (here I've used nn.Dense for a minimal reproducible example) using [flax.linen.vmap](https://flax.readthedocs.io/en/latest/_autosummary/flax.linen.vmap.html#flax.linen.vmap):

```
import jax
from flax import linen as nn
import jax.numpy as jnp


class Parallel(nn.Module):
    @nn.compact
    def __call__(self, x):
        BatchDense = nn.vmap(
            nn.Dense,
            axis_size     = 2,
            in_axes       = None,
            out_axes      = 0,
            variable_axes = {'params': 0},
            split_rngs    = {'params': True},
        )
        return BatchDense(4)(x)


model  = Parallel()
rng    = jax.random.PRNGKey(42)
x      = jnp.array([[1., 2., 3.]])
params = model.init(rng, x)
y      = model.apply(params, x)
```

How might I do something similar in Haiku?  Thanks in advance!

**Comment by dyth** [comment]
@tomhennigan sorry for tagging, but might there be an update?

**Comment by dyth** [comment]
I followed the discussion in https://github.com/google/jax/issues/7430 to come up with the following, but it seems quite hacky.  
```
import haiku as hk
import jax
import jax.numpy as jnp


def f(x):
    net = hk.nets.MLP([300, 100, 2])
    return net(x)

f = hk.transform(f)

ensemble_size = 4

rng  = jax.random.PRNGKey(42)
rngs = jax.random.split(rng, ensemble_size)
x    = jnp.ones([1, 28 * 28])

params = jax.vmap(lambda r: f.init(r, x))(rngs)
apply  = lambda params, rng, x: jax.vmap(lambda p: f.apply(p, rng, x), out_axes=1)(params)

x = jnp.ones([10, 28 * 28])
x = apply(params, None, x)
```
